### ISSUE TITLE
React router v4

### ISSUE THREAD

**Issue by tobyl** [react-boilerplate/react-boilerplate#1033]
I've been watching the rollout of react router v4 and it looks like a pretty exciting (but dramatic) departure from v2 best practices. Since the router is such a major part of the boilerplate I'm wondering what general opinion is on integrating it. It seems like this would be a major rewrite, but I'm curious what others are thinking about this.

**Comment by mxstbr** [comment]
Not sure yet, I haven't played with it at all. Looking forward to this discussion to hear what people think!

**Comment by amilajack** [comment]
I've only heard positive things about react-router 4. I think think we should eventually migrate to it despite it having breaking changes

**Comment by investlift** [comment]
React Router 4 is an attempt to re-simplify react js development.  It would be really nice if we start looking in that direction.

**Comment by sedubois** [comment]
Este recently [upgraded](https://github.com/este/este/pull/1156) to react-router 4.

**Comment by amilajack** [comment]
Is it stable already?

**Comment by peter-mouland** [comment]
I upgraded to React Router 4 and it was pleasant :)  The changes made to the system all make sense and actually enabled me to remove some of my own logic around routes.   The new setup seems more flexible to me.

this is the code changes I had to do which includes updating router tests: 
https://github.com/peter-mouland/react-lego/compare/universal...react-router-4

**Comment by sedubois** [comment]
@amilajack AFAIK, yes

**Comment by timdorr** [comment]
@sedubois No, it's not yet. We still have a number of things to figure out and that are in flux, all of which may affect your applications. I specifically marked the last version as alpha for this very reason. Please do not use it in your production applications yet.

**Comment by sedubois** [comment]
@timdorr sorry misunderstood the question, I only meant to say it worked in Este boilerplate.

**Comment by kachkaev** [comment]
Hey guys,

Has anyone given v4 a try in this boilerplate so far? I once used it in a CRA project and quite liked the new API! Not sure how easy it would be to integrate the new approach with the async routes in `react-boilerplate` though as `routes.js` and code splitting look a bit too complex to me.

If you've done any experiments with it, feel free to share a link to your fork in this thread! v4 is `alpha.6` as of today and their milestone progress looks quite promising:  
https://github.com/ReactTraining/react-router/milestone/18

**Comment by sammy6464** [comment]
@kachkaev I've just been looking over v4. I think the new API looks awesome too! Composing components directly seems so much more react-y rather than plugging react components into a separate API. 
The problem I see with integrating the current code splitting pattern is there doesn't seem to be an equivalent of the getComponent(nextState, cb) function which allows us to fulfil a promise asynchronously and then send the component, reducer and saga that are returned back via the callback.
For example the new Match component only takes these props: 
 const { children, render, component:Component, pattern } = this.props

So I'm thinking the code splitting (System.import, promise etc) would have to be applied downstream at the level of the component getting returned? Maybe there could be a higher order component that could deal with this? And all your 'top level' containers i.e the ones split up by routes could be wrapped in that HOC?

**Comment by peter-mouland** [comment]
@sammy6464 I've used react-router-4 with promises being resolved before rendering components, but had to switch my method to use the [matchRoutesToLocation](https://github.com/ReactTraining/react-router-addons-routes#api) from react-router-addons-routes package.

I've used in my react-lego app.  You can see here, if the component has *needs* (the component needs data before rendering) and the route matches then we fetch the data first and set the context. https://github.com/peter-mouland/react-lego/blob/koa-app/src/server/middleware/set-router-context.js#L4

**Comment by sammy6464** [comment]
@peter-mouland thanks a lot that looks really interesting! I'll digest it all over the Xmas break :) 
Yeah the HOC idea was a bit stupid, was just trying to think how you get something in between Match wanting to directly return a component and waiting on the promise. That's a cool approach though with the middleware and the data fetching!
I guess the moral of the story is it's not exactly a simple change, so here's about the time I slowly back away while all you much smarter people think of things hahaha.
Merry Xmas everyone!

**Comment by kachkaev** [comment]
v4 has been released! https://github.com/ReactTraining/react-router :tada: :tada: :tada: :tada:

**Comment by znafets** [comment]
I closed #1637 because it poses the same question ... now I am looking forward for a definite answer here

**Comment by Geczy** [comment]
Updates?

**Comment by jeanmichelcote** [comment]
Anybody working on this?

**Comment by vicbell** [comment]
Nothing happend?

**Comment by anuraaga** [comment]
Hello - I have set up a PR on my fork that replaces react-router in the boilerplate with v4. Can anyone look at it to see if the general design looks sane? If so, I'll update the code generators and docs as well and submit the PR to this repo.

https://github.com/anuraaga/react-boilerplate/pull/1/files

The main conceptual change is that the routes config is migrated to a routes component, and their render function is managed by a Bundle, which asynchronously loads the real component - the Bundle is taken basically as-is from the react-router documentation.

A less important change is moving the loading logic into the components themselves. This makes the route config much easier to read, but it's not really required, I can revert it if that'd be better.

**Comment by ax-hd** [comment]
Hello,
Any update on this? Will this boilerplate be updated with react-router 4? Any suggestions on how I can modify the current project to use react router 4?

**Comment by IamJoseph** [comment]
@anuraaga I copied your files and everything worked like a charm! Great work, what a time-saver. Only things I would change are 
1) installing and using prop-types instead of using react's since it's deprecated (But if you are focusing solely on react-router no biggie) 
2) remove react-router since you'll only need react-router-dom as a dependency

FYI people you may need to change App/index.js to a SFC or regular Component if yours is a PureComponent.

In my files I actually updated all dependencies to their latest. The only other issue I remember having were the image loaders in internals/webpack/webpack.babel.base.js

**Comment by anuraaga** [comment]
@iamjoseph Thanks for taking a look! I'll go ahead and update the PR to remove the unnecessary dependency and update the codegen and submit a PR here by early next week.

Incidentally, when integrating into a larger project, I did find perhaps the same issue of not rerendering if App is a redux connected component. It's because the reducer I slotted in bindly isn't immutable - fixing that helped, will update that part of the PR too.

**Comment by IamJoseph** [comment]
@anuraaga Have you figured out a way to load child components?

**Comment by anuraaga** [comment]
@IamJoseph child components should just involve writing a similar Switch statement in the parent component, with a similar structure as routes.js. Parent component routes can't be exact.

I think the status quo is preserved in that there's no full codegen for child routes, and the README could be updated with the new syntax. By generating loaders like the pattern I used in my PR, I think this will actually be a very small amount of syntax when writing child routes manually (or it could be copy-pasted out of routes.js which would be similar to the copy-pasting involved in the current workaround).

**Comment by IamJoseph** [comment]
I cannot tell if I'm more frustrated from having spent so long trying to nest child components instead of waiting for your response, or relieved that I eventually came to the same conclusion :) . Essentially the same thing I added 
```javascript
<AsyncRoute path={pathname} load={this.getSubRouteContent(pathname)} />
``` 
in the render and 
```javascript
getSubRouteContent = (path) => {
    switch (path) {
      case '/work/current_site':
        return currentLoader();
      case '/work/client_sites':
        return clientLoader();
      case '/work/projects':
        return projectsLoader();
      case '/work/graphics':
        return graphicsLoader();
      default:
      case '/work':
        return defaultLoader();
    }
  }
```
as a member

**Comment by anuraaga** [comment]
Sent #1746 to try to integrate this into the boilerplate. @IamJoseph you may be interested in the documentation update to look at the pattern I came up with for child routes.

https://github.com/anuraaga/react-boilerplate/blob/8123a1b8469e699d4fd15ac25d18287e5d1eea66/docs/js/routing.md

**Comment by justingreenberg** [comment]
closed via #1746
